src/DiagnosticsManager.ts

```diff
@@ -0,0 +1,291 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the VSCode Swift open source project
+//
+// Copyright (c) 2024 the VSCode Swift project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of VSCode Swift project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import * as vscode from "vscode";
+import stripAnsi = require("strip-ansi");
+import configuration from "./configuration";
+import { SwiftExecution } from "./tasks/SwiftExecution";
+import { WorkspaceContext } from "./WorkspaceContext";
+
+interface ParsedDiagnostic {
+    uri: string;
+    diagnostic: vscode.Diagnostic;
+}
+
+type DiagnosticsMap = Map<string, vscode.Diagnostic[]>;
+
+/**
+ * Handles the collection and deduplication of diagnostics from
+ * various {@link vscode.Diagnostic.source | Diagnostic sources}.
+ *
+ * Listens for running {@link SwiftExecution} tasks and allows
+ * external clients to call {@link handleDiagnostics} to provide
+ * thier own diagnostics.
+ */
+export class DiagnosticsManager implements vscode.Disposable {
+    // Prior to Swift 6 "sourcekitd" was the source
+    static sourcekit: string[] = ["SourceKit", "sourcekitd"];
+    static swiftc: string[] = ["swiftc"];
+
+    private diagnosticCollection: vscode.DiagnosticCollection =
+        vscode.languages.createDiagnosticCollection("swift");
+
+    constructor(context: WorkspaceContext) {
+        this.onDidStartTaskDisposible = vscode.tasks.onDidStartTask(event => {
+            // Will only try to provide diagnostics for `swift` tasks
+            const execution = event.execution.task.execution;
+            if (!(execution && execution instanceof SwiftExecution)) {
+                return;
+            }
+            // Clean up old "swiftc" diagnostics
+            this.removeBuildDiagnostics();
```

Comment:
Prior to this PR build errors were cleared after a build completed and before the new build's errors were added to the problems view. This meant you had the list of build errors from the previous build available while you were in the process of doing a new build.

Reply:
done